src/main/java/com/limechain/trie/structure/decoded/node/DecodedNode.java

```diff
@@ -228,4 +240,221 @@ public byte[] calculateMerkleValue(Function<byte[], byte[]> hashFunction, boolea
+     * @return DecodedNode<Nibbles, List < Byte>> The decoded node with its partial key, children, and storage value.
+     * @throws NodeDecodingException If the encoded array is null, empty, or invalid.
+     * @throws IllegalArgumentException If 'encoded' is null.
+     */
+    public static DecodedNode decode(byte[] encoded) {
+        ScaleCodecReader reader = new ScaleCodecReader(encoded);
+        if (encoded == null || encoded.length == 0) {
+            throw new NodeDecodingException("Invalid node value: it's empty");
+        }
+
+        int firstByte = reader.readByte() & 0xFF;
+        boolean hasChildren;
+        Boolean storageValueHashed;
+        int pkLenFirstByteBits;
+
+        //https://spec.polkadot.network/#defn-node-header
+        switch (firstByte >> 6) {
+            case 0b00:
+                if ((firstByte >> 5) == 0b001) {
+                    hasChildren = false;
+                    storageValueHashed = true;
+                    pkLenFirstByteBits = 5;
+                } else if ((firstByte >> 4) == 0b0001) {
+                    hasChildren = true;
+                    storageValueHashed = true;
+                    pkLenFirstByteBits = 4;
+                } else if (firstByte == 0) {
+                    hasChildren = false;
+                    storageValueHashed = null;
+                    pkLenFirstByteBits = 6;
+                } else {
+                    throw new NodeDecodingException("Invalid header bits");
+                }
+                break;
+            case 0b10:
+                hasChildren = true;
+                storageValueHashed = null;
+                pkLenFirstByteBits = 6;
+                break;
+            case 0b01:
+                hasChildren = false;
+                storageValueHashed = false;
+                pkLenFirstByteBits = 6;
+                break;
+            case 0b11:
+                hasChildren = true;
+                storageValueHashed = false;
+                pkLenFirstByteBits = 6;
+                break;
+            default:
+                throw new NodeDecodingException("Unknown variant");
+        }
+
+
+        int pkLen = getPkLen(reader, firstByte, pkLenFirstByteBits);
+
+        if (pkLen != 0 && !hasChildren && storageValueHashed == null) {
+            throw new NodeDecodingException("Empty trie with partial key ");
+        }
+
+        byte[] partialKeyLEBytes = extractPartialKey(reader, pkLen);
+
+        int childrenBitmap = extractChildrenBitmap(reader, hasChildren);
+
+        byte[] storageValue = extractStorageValue(reader, storageValueHashed);
+
+        List<List<Byte>> children = extractChildren(reader, childrenBitmap);
+        Nibbles partialKey = Nibbles.fromLEBytes(partialKeyLEBytes);
+        if (pkLen % 2 == 1) {
+            /// This is for situations where the partial key contains a `0` prefix that exists for
+            /// alignment but doesn't actually represent a nibble.
```

Comment:
Please use consistent comment style
```suggestion
            // This is for situations where the partial key contains a `0` prefix that exists for
            // alignment but doesn't actually represent a nibble.
```